### PR TITLE
Support all kinds of generated vala and vapi sources

### PR DESCRIPTION
**Pull request by nirbheek**
This is the first step in making Vala support have feature-parity with C/C++ support. Vala and Vapi sources generated with Generators and CustomTargets are no longer ignored. Dependencies are setup properly and they are added to the commandline.

Also does some cleanup of how generated sources are handled while generating C/C++ targets.

Includes two tests: one for generated Vala sources, and one for generated C sources in Vala targets.

This lays groundwork needed for #895 which uses CustomTargets for generating vapis.

### CODE REVIEW

**Comment by nirbheek** [comment]
Forgot to mention, this depends on #866, so I've included the commits from that in here for now.

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -1590,6 +1605,9 @@ def get_link_debugfile_args(self, linker, target, outname):
         return linker.get_link_debugfile_args(outname)
 
     def generate_single_compile(self, target, outfile, src, is_generated=False, header_deps=[], order_deps=[]):
+        """
+        Compiles only C/C++ sources
```
Comment (jpakkane) [inline mesonbuild/backend/ninjabackend.py]: Also Objective C and Objective C++

Comment (nirbheek) [inline mesonbuild/backend/ninjabackend.py]: Fixed!

**Comment by nirbheek** [comment]
Rebased!

**Pull request merged**